prima implementazione sonificazione

knickita opened the pull request:
usare SPAZIO per attivare e disattivare la sonificazion

knickita committed:
prima implementazione sonificazione

usare SPAZIO per attivare e disattivare la sonificazion

Mattia541993 commented:
Prima le cose più di convenzione: per le funzioni, variabili e inject va bene se facciamo che iniziano con la lettera minuscola (tipo `ScaleFunctionInY`)?.

Per la parte di sonificazione: vedo molto bene adottare `Transport` per la sonificazione batch (0D). Mi piace anche l'utilizzo del `Channel`. Ho visto però che hai tolto tutta la parte di tipi di sonificazione e hai messo come parametri globali parametri come la frequenza che secondo me è giusto che restino per strumento. Inoltre con un numero di campioni la sonificazione continua non la si può implementare perchè comunque genera sempre 32 campioni nella tua pull request. In generale: come mai hai modificato così tanto anche la struttura generale del file? Non ti tornava qualcosa sulla modellazione dei singoli strumenti?

Mattia541993 commented:
Poi ho visto che hai toccato il componente `FunctionPlot` per fare le prove. Va bene, ma considera che quel componente poi non sarà più lì nella struttura del sito. Anzi, probabilmente essendo il file che arriva dal progetto di prova verrà proprio tolto, quindi per fare dei test puoi continuare lì. Poi appena riesco creo la struttura anche delle altre pagine in modo che tu possa lavorare direttamente da lì

knickita commented:
- io come regola terrei che le variabili iniziano con la minuscola e le funzioni iniziano con la maiuscola. così (soprattutto nelle inject) è sempre chiaro se è una variabile o una funzione
- per la sonificazione sono ripartito da zero col file solo per fare un pò di chiarezza a me stesso su che giro dovessero fare le varie cose. sto per fare un'altro commit su quel branch dove reintroduco i due strumenti che avevi configurato tu (flauto e chitarra). 
- hai ragione sul discorso delle frequenze, mi ero portato i parametri da cambiare sulle prime righe perchè ho fatto un po di esperimenti sulla resa acustica e sulla piacevolezza del suono, ma non serve per forza che restino globali. non li porterei dentro gli strumenti però, perchè tanto dobbiamo trovare un range di frequenze su cui vogliamo lavorare, e poi gli strumenti si accorderanno tutti su quelle frequenze (ad esempio oggi siamo tra 110hz e 880hz, che sono frequenze sulla medio bassa che di solito trasmettono calore, e mi sembravano piacevoli. in generale non mi spingerei oltre ai 1000hz, perchè poi tra 1000hz e 4000hz ci sono le timbriche della voce che rendono intellegibile il parlato. nell'ottica di un TTS lascerei quel renge di frequenze pulito per permettere alle parole di essere comprese mentre sta suonando la melodia.
- i 32 campioni ovviamente non funzionano nella sonificazione continua, che sto finendo di implementare. in realtà non funziona neanche il concetto di Transport se deve essere veramente continua bisogna implementare qualcosa che aggiorni 30 o 60 volte al secondo, non basandosi su quanto tempo deve suonare una nota.
- in generale ho modificato la struttura perchè sono ripartito un pò da zero cercando di capire bene come funzionasse il discorso di sound.js. mi aspetto nel prossimo push di avere come interfaccia esterna a quel file le seguenti funzioni:
StartAudio, StopAudio, IsAudioStarted, SetContinuousMode, SetDiscreteMode
per ora terrei solo i due strumenti già pensati (chitarra per la DiscreteMode e flauto per la ContinuousMode. capiamo poi più avanti se inserire altri strumenti.. ad essere sincero ho provato a giocare un pò con i parametri della chitarra ma la trovo un pò troppo "sporca elettricamente" e poco confortevole.. mi piace di più il synth che ho configurato, però bisognerebbe anche fare un'oretta con Cristian secondo me, dove presentiamo un pò di ipotesi e ci da anche lui il suo parere sul feeling.
- FunctionPlot l'ho toccato solo perchè dovevo inserire un tasto per fare la sonificazione, so che verrà rimosso. l'unica cosa da tenere in considerazione è che i suoni non possono partire in automatica su tutti i browser, ma per motivi di protocolli alcuni browser richiedono che venga attivata una callback da un azione dell'utente (bottono html premuto, o tasto fisico sulla tastiera). quindi sarà comunque da prevedere nell'interfaccia un bottone che attivi/disattivi l'audio, con relativa shortcut (per ora è lo spazio ma ovviamente poi decidiamo)

knickita committed:
piccole aggiunte alla sonificazione

reinserite chitarra e clarinetto, minor bugfix nella funzione ScaleFunctionInY, dove si rischiava (in un caso limite) di andare fuori dalla lunghezza dell'array delle note disponibili..

knickita commented:
ho fatto un push su questa branch reinserendo chitarra e clarinetto per poter fare dei confronti.
alla riga 69 di sound.js puoi mettere "guitar" o "clarinet" al posto di "synth" così senti anche tu le differenze..
a me la chitarra non sta piacendo, magari provo a capire se c'è un'altro modo a livello di sintesi sonora per rappresentarla..

Mattia541993 commented:
- sulle inject va bene come dici tu
- non sono d'accordo sul range delle frequenze globali. Ogni strumento ha le sue frequenze sulle quali funziona meglio e una persona che ascolta un certo strumento è abituato a sentirlo per un certo range di frequenze. Se anche stiamo sui 1000Hz massimi gli armonici si vanno comunque a sovrapporre col parlato. La massima sensibilità umana la si ha intorno ai 1000Hz. Lì ce la possiamo giocare coi volumi. Comunque non stiamo parlando di frequenze fondamentali che vanno verso un campo delle alte frequenze e quindi che possono causare fastidio o non essere udite. 
- La sonificazione continua per com'era implementata con il `triggerRelease` seguito da un `triggerAttack` non andava bene?
- sono d'accordo con i metodi di interfaccia e anche sul fatto che la chitarra se si riuscisse a migliorare o trovare qualcosa di più naturale sarebbe molto bello. Anche il timbro del tuo synth mi piace
- sì, per `FunctionPlot` te l'ho detto giusto per ricordartelo.. Devi avere alla fine un posto in cui fare le prove. 
- Per l'audio attivato da un'azione esplicita dell'utente se anche il browser supportasse lo start automatico lo disabiliterei per questioni di accessibilità.

Ora guardo il commit
